rectools/visuals/visual_app.py

```diff
@@ -144,20 +195,153 @@ def _group_recos(
+            }
+        )
+        return df
+
+    def save(self, folder_name: str, overwrite: bool = False) -> None:
+        """Save stored data for `VisualApp` widgets. This method is not meant to be used
+        directly. Use `VisualApp` or `ItemToItemVisualApp` class methods instead.
+
+        Parameters
+        ----------
+        folder_name : str
+            Destination folder for data.
+        overwrite : bool, default ``False``
+            Allow to overwrite in the folder files if they already exist.
+        """
+        interactions_df = self._ungroup_interactions(
+            grouped_interactions=self.grouped_interactions, selected_requests=self.selected_requests, id_col=self.id_col
+        )
+        recos_df = self._ungroup_recos(
+            grouped_recos=self.grouped_recos, selected_requests=self.selected_requests, id_col=self.id_col
+        )
+        requests_df = self._create_requests_df(self.selected_requests)
+
+        Path(folder_name).mkdir(parents=True, exist_ok=True)
+        mode = "w" if overwrite else "x"
+        interactions_df.to_csv(Path(folder_name, StorageFiles.Interactions), index=False, mode=mode)
+        recos_df.to_csv(Path(folder_name, StorageFiles.Recos), index=False, mode=mode)
+        requests_df.to_csv(Path(folder_name, StorageFiles.Requests), index=False, mode=mode)
+
+    @classmethod
+    def from_saved(cls, folder_name: str) -> "AppDataStorage":
+        r"""Load prepared data for VisualApp widgets. This method is not meant to be used
+        directly. Use `VisualApp` or `ItemToItemVisualApp` class methods instead.
+
+        Parameters
+        ----------
+        folder_name : str
+            Folder where data was saved earlier.
+
+        Returns
+        -------
+        AppDataStorage
+            Data storage class for visualisation widgets.
+        """
+        interactions = pd.read_csv(Path(folder_name, StorageFiles.Interactions))
+        recos = pd.read_csv(Path(folder_name, StorageFiles.Recos))
+        selected_requests = dict(
+            pd.read_csv(
+                Path(folder_name, StorageFiles.Requests),
+            ).values
+        )
+
+        if Columns.TargetItem in interactions.columns and Columns.User in interactions.columns:
+            raise ValueError(
+                """Unable to create VisualApp. Saved interactions have both columns:
+                {Columns.TargetItem} and {Columns.User}"""
+            )
+
+        if Columns.User in interactions.columns:
+            is_u2i = True
+            id_col = Columns.User
+        elif Columns.TargetItem in interactions.columns:
+            is_u2i = False
+            id_col = Columns.TargetItem
+        else:
+            raise ValueError(
+                """Unable to create VisualApp. Saved interactions don't have any of the columns:
+                {Columns.TargetItem} or {Columns.User}"""
+            )
+
+        dummy_item_data = pd.DataFrame(columns=[Columns.Item])
```

Comment:
I don't really like this and you know why :)

Reply:
yep

Reply:
fixed